wp/lib/bap_wp/src/constraint.ml

```diff
@@ -222,18 +222,72 @@ let rec eval_aux ?stats:(stats = init_stats) (constr : t) (olds : z3_expr list)
       (eval_aux ~stats:stats c2 olds news ctx)
   | Clause (hyps, concs) ->
     let eval_conjunction conj =
-      List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
-      |> Bool.mk_and ctx
+      if List.length conj = 1 then 
+        (* This is tail recursive. Avoids And node. *)
+        eval_aux ~stats:stats (List.hd_exn conj) olds news ctx 
+      else
+        List.map conj ~f:(fun c -> eval_aux ~stats:stats c olds news ctx)
+        |> Bool.mk_and ctx
     in
-    let concs_expr = eval_conjunction concs in
     if List.is_empty hyps then
-      concs_expr
+      eval_conjunction concs(* possibly tail recursive? *)
     else
+      let concs_expr = eval_conjunction concs in
       let hyps_expr = eval_conjunction hyps in
       Bool.mk_implies ctx hyps_expr concs_expr
   | Subst (c, o, n) ->
+    (* The following should be functionally equivalent to 
+       [eval_aux ~stats:stats c (olds @ o) (news @ n') ctx] with changes made to achieve 
+       better performance. Duplicate keys are pruned from the olds list.
+
+       NOTE : It is assumed as an invariant that the olds list does not contain duplicate 
+       keys. Failure to maintain this invariant may result in incorrect behavior.
+
+       Using a Map data structure was tried and it was found that shuffling in and out to 
+       lists for Expr.substitute made it slow. Instead, two lists are maintained and 
+       pruned of duplicate keys. The most recently used items are put at the front of the
+       list to make it fast to find the most commonly used variables.
+    *)
     let n' = List.map n ~f:(fun x -> Expr.substitute x olds news) in
-    eval_aux ~stats:stats c (olds @ o) (news @ n') ctx
+    if (List.length n') = 1 (* It appears we generate overwhelmingly size 1 lists *)
+    then 
+      begin
+        let o = List.hd_exn o in
+        let n' = List.hd_exn n' in
+        let findn e l =
+          (* [findn' n e l] finds element [e] in list [l] while maintaining an 
+             accumulator of the current index [n] *)
+          let rec findn' n e l = 
```

Comment:
Can we just use `List.findi`? Feels like we're re-inventing the wheel.

Reply:
Where is such a function

Reply:
Oh I found it. nice.